internal/implementations/implementation.go

```diff
@@ -0,0 +1,94 @@
+package implementations
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/events"
+	"github.com/nginxinc/nginx-kubernetes-gateway/pkg/sdk"
+)
+
+type ImplementationImpl struct {
+	logger       logr.Logger
+	objectType   client.Object
+	eventCh      chan<- interface{}
+	filter       NamespacedNameFilter
+	resourceKind string
+}
+
+var _ sdk.Implementation = &ImplementationImpl{}
+
+type NamespacedNameFilter func(nsname types.NamespacedName) (bool, string)
+
+func NewImplementation(objectType client.Object, logger logr.Logger, eventCh chan<- interface{}) *ImplementationImpl {
+	return NewImplementationWithFilter(objectType, logger, eventCh, nil)
+}
+
+func NewImplementationWithFilter(
+	objectType client.Object,
+	logger logr.Logger,
+	eventCh chan<- interface{},
+	filter NamespacedNameFilter,
+) *ImplementationImpl {
+	return &ImplementationImpl{
+		objectType:   objectType,
+		logger:       logger,
+		eventCh:      eventCh,
+		filter:       filter,
+		resourceKind: reflect.TypeOf(objectType).Elem().Name(),
+	}
+}
+
+func (impl *ImplementationImpl) Upsert(obj client.Object) {
+	t := reflect.TypeOf(impl.objectType)
```

Comment:
Why not just attempt a type cast here?

Reply:
not sure I fully understand the suggestion. Could you share a snippet?